src/composables/useEventStore/index.ts

```diff
@@ -0,0 +1,101 @@
+import { computed, ref, shallowRef } from 'vue'
+import type { TresContext } from '../useTresContextProvider'
+import { useRaycaster } from '../useRaycaster'
+import { createGlobalState } from '@vueuse/core'
+import { Intersection, Event, Object3D } from 'three'
+import { hyphenate } from '../../utils'
+
+export const useEventStore = createGlobalState(
+  (scene, context, emit) => {
+    const _scene = shallowRef<THREE.Scene>()
+    const _context = shallowRef<TresContext>()
+
+    if(scene) _scene.value = scene
+    if(context) _context.value = context
+
+    const sceneChildren = computed(() => {
+      return _scene.value ? _scene.value.children : []
+    })
+
+    function executeEventListeners(listeners: Function | Function[], intersection, event){
+      // Components with multiple event listeners will have an array of functions
+      if (Array.isArray(listeners)) {
+        for (const listener of listeners) {
+          listener(intersection, event);
+        }
+      }
+
+      // Single listener will be a function
+      if (typeof listeners === 'function') {
+        listeners(intersection, event);
+      }
+    }
+
+    /**
+     * propogateEvent
+     * 
+     * Propogates an event to all intersected objects and their parents
+     * @param eventName - The name of the event to propogate
+     * @param event - The event object
+     * @param intersects - An array of intersections
+     */
+    function propogateEvent(eventName: string, event, intersects) {
+      let stopPropagating = false;
+      
+      // Loop through all intersected objects and call their event handler
+      if (intersects.length) {
+        for(const intersection of intersects) {
+          if (stopPropagating) return;
+          intersection.stopPropagation = () => stopPropagating = true;
+
+          const { object } = intersection
+          executeEventListeners(object[eventName], intersection, event)
+
+          // Propogate the event up the parent chain before moving on to the next intersected object
+          let parent = object.parent
+          while(parent !== null && !stopPropagating) {
+            executeEventListeners(parent[eventName], intersection, event)
+            parent = parent.parent
```

Comment:
## One click should probably be no more than one call to a click handler.

For a given click, etc., a particular click handler shouldn't be called more than once, it seems to me.

## Currently

This shows the result of a single click with the current bubbling setup:

<img width="1285" alt="Screenshot 2024-02-02 at 02 42 39" src="https://github.com/Tresjs/tres/assets/20469369/5476e9d2-50eb-499e-9939-988e7a3f3965">

The console shows that the box at the top of the pyramid gets its click handler called 4 times.

## "Realer" world

The above doesn't seem wrong to me, but I was trying to think about it as a Tres user and I think it breaks down in "realer" world cases.

Here's a simple example: buy a bicycle.

```vue
<Bicycle @click="buy">
  <Frame>
    <Wheel />
    <Wheel />
  </Frame>
</Bicycle>
```

Positioning the canvas at a certain angle, you could click both wheels and the frame, and you'll buy 3 bikes instead of 1.

## R3F removes duplicates

I checked out R3F. [They hey remove duplicates.](https://github.com/pmndrs/react-three-fiber/blob/0ef66a1d23bf16ecd457dde92b0517ceec9861c5/packages/fiber/src/core/events.ts#L191C5-L191C41) 

Here's a [StackBlitz with a simple setup like your playground](https://stackblitz.com/edit/vitejs-vite-ggmpvs?file=src%2FApp.tsx,src%2Fcomponents%2FBox.tsx).

Reply:
Great catch!

I will be sure to add duplicate checks somehow

Reply:
The latest changes include duplicate checks and it's working pretty well

Thanks for catching that!